src/kernel/arm/cm4/process.c

```diff
@@ -0,0 +1,95 @@
+#include <kernel/process.h>
+#include <stdbool.h>
+#include <multitasking.h>
+/**
+ * Fake an exception entry frame for a process, which makes starting a new
+ * process identical to switching to it.  A context switch stack push is also
+ * faked here, to make it even easier to start the process.
+ * @param proc PCB whose stack should be initialized.
+ * @param uses_fp whether the process will use the floating point unit or not.
+ * @param kernel_mode whether the process will run in kernel mode or not.
+ * @return 0 on success, -1 on failure.
+ */
+int process_init(pcb_t *proc, int (*main)(void),
+        uint32_t *stack_start, size_t stack_size,
+        bool uses_fp, bool kernel_mode) {
+    int framesize = 0x20;
+    uint32_t spmask = 0;
+    bool forcealign = SCB->CCR & SCB_CCR_STKALIGN_Msk;
+
+    proc->main = main;
+    proc->stack_lim = (uint32_t)stack_start;
+    proc->stack_top = (uint32_t)(stack_start + stack_size);
+    proc->stack_ptr = proc->stack_top;
+    proc->pid = 1;
+
+    if(uses_fp) {
+        framesize = 0x68;
```

Comment:
Some of the numbers could be made into defines but I'm not sure if its worth the time

Reply:
There is a reason for this - the code looks like the pseudocode in the ARMv7-M Technical Reference Manual, Section B 1.5.6: Behavior on Exception Entry. (please help me)
I wanted it to be easy to recognize.  I will add a comment block saying where to find the constants in the manual.